addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/internal/CupsHandlerFactory.java

```diff
@@ -0,0 +1,114 @@
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+
+/**
+ * The {@link CupsHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ * 
+ * @author Tobias Braeutigam - Initial contribution
+ */
+public class CupsHandlerFactory extends BaseThingHandlerFactory {
+	private Logger logger = LoggerFactory.getLogger(CupsHandlerFactory.class);
+	
+	private ServiceRegistration<?> discoveryServiceReg;
+
+	@Override
+	public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+		return CupsBindingConstants.SUPPORTED_TYPES_UIDS.contains(thingTypeUID);
+	}
+
+	@Override
+	public Thing createThing(ThingTypeUID thingTypeUID,
+			Configuration configuration, ThingUID thingUID, ThingUID bridgeUID) {
+		logger.info("createThing(" + thingTypeUID + "," + configuration + ","
+				+ thingUID + "," + bridgeUID + ")");
+		if (CupsBindingConstants.SUPPORTED_BRIDGE_TYPES_UIDS
+				.contains(thingTypeUID)) {
+			logger.info("creating bridge {}", thingTypeUID);
+			return super.createThing(thingTypeUID, configuration, thingUID,
+					null);
+		}
+		if (CupsBindingConstants.SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID)) {
+			ThingUID deviceUID = getCupsPrinterUID(thingTypeUID, thingUID,
+					configuration, bridgeUID);
+			logger.info("creating thing " + thingTypeUID
+					+ " from deviceBridgeUID: " + deviceUID + " Bridge: "
+					+ bridgeUID);
+			return super.createThing(thingTypeUID, configuration, deviceUID,
+					bridgeUID);
+		}
+		throw new IllegalArgumentException("The thing type " + thingTypeUID
+				+ " is not supported by the binding.");
+	}
+	
+	private ThingUID getCupsPrinterUID(ThingTypeUID thingTypeUID, ThingUID thingUID, Configuration configuration,
+			 ThingUID bridgeUID) {
+		if (thingUID == null) {
+			 String name = (String) configuration.get(CupsBindingConstants.PRINTER_PARAMETER_URL);
+			 thingUID = new ThingUID(thingTypeUID, name, bridgeUID.getId());
+		}
+		return thingUID;
+	}
+
+	@Override
+	protected ThingHandler createHandler(Thing thing) {
+
+		ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+		if (thingTypeUID.equals(PRINTER_THING_TYPE)) {
+			return new CupsHandler(thing);
+		} else if (thingTypeUID.equals(SERVER_BRIGDE_THING_TYPE)) {
+			CupsBridgeHandler bridge = new CupsBridgeHandler((Bridge) thing);
+			registerDeviceDiscoveryService(bridge);
+			return bridge;
+		}
+
+		return null;
+	}
+
+	private void registerDeviceDiscoveryService(CupsBridgeHandler bridge) {
+		CupsDeviceDiscoveryService discoveryService = new CupsDeviceDiscoveryService();
+		discoveryService.activate(bridge);
+		this.discoveryServiceReg = bundleContext.registerService(
```

Comment:
There should be one such discoveryServiceReg per bridge and not just one in total, right? So better keep them in a Map instead.